File: backend/app/services/extraction.py
"""PDF, DOCX, and TXT extraction with page-aware source sections."""

from __future__ import annotations

import io
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import fitz
from docx import Document as DocxDocument

from app.exceptions import ValidationError


@dataclass(frozen=True, slots=True)
class ExtractedSection:
    """One extraction unit retaining source location."""

    text: str
    page_number: int | None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True, slots=True)
class ExtractionResult:
    """Normalized extraction output."""

    sections: list[ExtractedSection]
    page_count: int | None
    metadata: dict[str, Any]


def extract_pdf(content: bytes) -> ExtractionResult:
    """Extract text and metadata from every PDF page using PyMuPDF."""
    try:
        with fitz.open(stream=content, filetype="pdf") as pdf:
            metadata = {key: value for key, value in (pdf.metadata or {}).items() if value}
            sections = [
                ExtractedSection(
                    text=page.get_text("text").strip(),
                    page_number=index + 1,
                    metadata={"page": index + 1},
                )
                for index, page in enumerate(pdf)
                if page.get_text("text").strip()
            ]
            return ExtractionResult(
                sections=sections,
                page_count=pdf.page_count,
                metadata=metadata,
            )
    except Exception as exc:
        raise ValidationError("PDF text extraction failed.") from exc


def extract_docx(content: bytes) -> ExtractionResult:
    """Extract headings, paragraphs, and core properties from DOCX."""
    try:
        document = DocxDocument(io.BytesIO(content))
    except Exception as exc:
        raise ValidationError("DOCX text extraction failed.") from exc
    sections: list[ExtractedSection] = []
    current_heading: str | None = None
    paragraph_buffer: list[str] = []

    def flush() -> None:
        if paragraph_buffer:
            text = "\n".join(paragraph_buffer).strip()
            if text:
                sections.append(
                    ExtractedSection(
                        text=text,
                        page_number=None,
                        metadata={"heading": current_heading} if current_heading else {},
                    )
                )
            paragraph_buffer.clear()

    for paragraph in document.paragraphs:
        text = paragraph.text.strip()
        if not text:
            continue
        if paragraph.style and paragraph.style.name.lower().startswith("heading"):
            flush()
            current_heading = text
            paragraph_buffer.append(text)
        else:
            paragraph_buffer.append(text)
    flush()
    properties = document.core_properties
    metadata = {
        key: value
        for key, value in {
            "title": properties.title,
            "subject": properties.subject,
            "author": properties.author,
            "keywords": properties.keywords,
            "category": properties.category,
        }.items()
        if value
    }
    return ExtractionResult(sections=sections, page_count=None, metadata=metadata)


def extract_txt(content: bytes) -> ExtractionResult:
    """Decode UTF-8 plain text."""
    text = content.decode("utf-8-sig").strip()
    return ExtractionResult(
        sections=[ExtractedSection(text=text, page_number=None)] if text else [],
        page_count=None,
        metadata={},
    )


def extract_document(filename: str, content: bytes) -> ExtractionResult:
    """Route extraction by the already validated extension."""
    extension = Path(filename).suffix.lower()
    if extension == ".pdf":
        result = extract_pdf(content)
    elif extension == ".docx":
        result = extract_docx(content)
    elif extension == ".txt":
        result = extract_txt(content)
    else:
        raise ValidationError("Unsupported document type.")
    if not result.sections:
        raise ValidationError(
            "No extractable text was found. Scanned PDFs require a separate OCR pipeline."
        )
    return result

File: backend/src/services/extraction.js
import path from "node:path";
import mammoth from "mammoth";
import { getDocument } from "pdfjs-dist/legacy/build/pdf.mjs";
import { ValidationError } from "../errors.js";
import { logProcessFinished, logProcessStarted } from "../logger.js";

function decodeHtml(value) {
  return value
    .replaceAll("&nbsp;", " ")
    .replaceAll("&amp;", "&")
    .replaceAll("&lt;", "<")
    .replaceAll("&gt;", ">")
    .replaceAll("&quot;", '"')
    .replaceAll("&#39;", "'");
}

function stripTags(value) {
  return decodeHtml(value.replace(/<[^>]*>/g, " ").replace(/\s+/g, " ").trim());
}

export async function extractPdf(content) {
  logProcessStarted("Extract PDF text", { bytes: content.length });
  try {
    const document = await getDocument({
      data: new Uint8Array(content),
      disableWorker: true,
      useSystemFonts: true,
    }).promise;
    const sections = [];
    for (let pageNumber = 1; pageNumber <= document.numPages; pageNumber += 1) {
      const page = await document.getPage(pageNumber);
      const textContent = await page.getTextContent();
      const text = textContent.items
        .map((item) => ("str" in item ? item.str : ""))
        .join(" ")
        .replace(/\s+/g, " ")
        .trim();
      if (text) {
        sections.push({ text, pageNumber, metadata: { page: pageNumber } });
      }
    }
    const pageCount = document.numPages;
    const metadataResult = await document.getMetadata().catch(() => ({}));
    const metadata = Object.fromEntries(
      Object.entries(metadataResult.info ?? {}).filter(
        ([, value]) => ["string", "number", "boolean"].includes(typeof value) && value !== "",
      ),
    );
    await document.destroy();
    logProcessFinished("Extract PDF text", {
      pages: pageCount,
      sections: sections.length,
      metadata,
    });
    return { sections, pageCount, metadata };
  } catch (error) {
    throw new ValidationError("PDF text extraction failed.", { cause: error });
  }
}

export async function extractDocx(content) {
  logProcessStarted("Extract DOCX text", { bytes: content.length });
  try {
    const { value: html } = await mammoth.convertToHtml({ buffer: content });
    const sections = [];
    let currentHeading = null;
    let buffer = [];
    const flush = () => {
      const text = buffer.join("\n").trim();
      if (text) {
        sections.push({
          text,
          pageNumber: null,
          metadata: currentHeading ? { heading: currentHeading } : {},
        });
      }
      buffer = [];
    };

    for (const match of html.matchAll(/<(h[1-6]|p|li)[^>]*>([\s\S]*?)<\/\1>/gi)) {
      const tag = match[1].toLowerCase();
      const text = stripTags(match[2]);
      if (!text) {
        continue;
      }
      if (tag.startsWith("h")) {
        flush();
        currentHeading = text;
      }
      buffer.push(text);
    }
    flush();

    if (sections.length === 0) {
      const { value } = await mammoth.extractRawText({ buffer: content });
      if (value.trim()) {
        sections.push({ text: value.trim(), pageNumber: null, metadata: {} });
      }
    }
    const result = { sections, pageCount: null, metadata: {} };
    logProcessFinished("Extract DOCX text", { sections: sections.length });
    return result;
  } catch (error) {
    throw new ValidationError("DOCX text extraction failed.", { cause: error });
  }
}

export function extractTxt(content) {
  logProcessStarted("Extract TXT text", { bytes: content.length });
  const text = new TextDecoder("utf-8", { fatal: true })
    .decode(content)
    .replace(/^\uFEFF/, "")
    .trim();
  const result = {
    sections: text ? [{ text, pageNumber: null, metadata: {} }] : [],
    pageCount: null,
    metadata: {},
  };
  logProcessFinished("Extract TXT text", {
    sections: result.sections.length,
    chars: text.length,
  });
  return result;
}

export async function extractDocument(filename, content) {
  logProcessStarted("Extract document", { filename, bytes: content.length });
  const extension = path.extname(filename).toLowerCase();
  let result;
  if (extension === ".pdf") {
    result = await extractPdf(content);
  } else if (extension === ".docx") {
    result = await extractDocx(content);
  } else if (extension === ".txt") {
    result = extractTxt(content);
  } else {
    throw new ValidationError("Unsupported document type.");
  }
  if (result.sections.length === 0) {
    throw new ValidationError(
      "No extractable text was found. Scanned PDFs require a separate OCR pipeline.",
    );
  }
  logProcessFinished("Extract document", {
    filename,
    extension,
    sections: result.sections.length,
    page_count: result.pageCount,
  });
  return result;
}
